### PR TITLE
Add Implemented BananaQueue With Tests And Documentation

### PR DESCRIPTION
**Pull request by ike-elue**
**Add**
- BananaQueue and relevant classes to util folder
- BananaQueueTest.java, tests operations that change BananaQueue's internal state

### CODE REVIEW

**Review by GabrielGonzalezDiaz — APPROVED** [review]

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.*;
+
+/* Nodes in the BananaQueue. Holds references to PeelQueues. */
+public class BananaNode {
+  private boolean complete;
+  
+  /* Uses Linked List model. */
+  private BananaNode prevNode, nextNode;
+
+  private HashMap<String, PeelQueue> peels;
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java:27]: add a comment to describe the map key and value

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  /**
+   * Returns true if String mapping for PeelQueue exists.
+   * 
+   * @param peelQueueTag Tag to associated PeelQueue.
+   * @return if String mapping for PeelQueue exists.
+   */
+  public boolean peelQueueExists(String peelQueueTag) {
+    return peels.containsKey(peelQueueTag);
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java]: toLowerCase()?

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  /**
+   * Returns PeelQueue given its tag.
+   * 
+   * @param peelQueueTag Tag to associated PeelQueue.
+   * @return PeelQueue mapped to tag
+   */
+  public PeelQueue getPeelQueue(String peelQueueTag) {
+    return peels.get(peelQueueTag);
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java]: toLowerCase()?

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.util;
+
+import java.util.*;
+
+/* Nodes in the BananaQueue. Holds references to PeelQueues. */
+public class BananaNode {
+  private boolean complete;
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java:21]: add a comment

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  /**
+   * Enqueues given PeelNode to PeelQueue associated with tag if PeelNode has a value and the BananaQueue's front element
+   * has a PeelQueue with the same associated name.
+   *
+   * @param peelQueueTag Tag to associated PeelQueue.
+   * @param peel PeelNode to add.
+   */
+  public void enqueuePeel(String peelQueueTag, PeelNode peel) {
+    if(peel == null || head == null || !head.peelQueueExists(peelQueueTag)) {
+      return;
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: better to return a boolean to indicate if the operation is successful

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    peels.put(peelQueueTag.toLowerCase(), peelQueue);
+  }
+
+  /**
+   * Removes String mapping to stored PeelQueue given a tag. Converts tag to lowercase.
+   * 
+   * @param peelQueueTag Tag to associated PeelQueue.
+   */
+  public void removePeelQueue(String peelQueueTag) {
+    peels.remove(peelQueueTag.toLowerCase());
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java]: better return a boolean to indicate if the operation is successful

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+    head.getPeelQueue(peelQueueTag).enqueue(peel);
+  }
+
+  /**
+   * Dequeues PeelQueue associated with the tag if the BananaQueue's front element has a PeelQueue with the same associated name.
+   *
+   * @param peelQueueTag Tag to associated PeelQueue.
+   */
+  public void dequeuePeel(String peelQueueTag) {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: better to return the removed PeelNode

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+    head.getPeelQueue(peelQueueTag).dequeue();
+  }
+
+  /**
+   * Adds generic PeelQueue to the front element of the BananaQueue with the given tag. 
+   *
+   * @param peelQueueTag Tag to associated PeelQueue.
+   */
+  public void addPeelQueue(String peelQueueTag) {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: what's the use case of adding an empty PeelQueue

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    addPeelQueue(peelQueueTag, new PeelQueue());
+  }
+  
+  /**
+   * Adds given PeelQueue to the front element of the BananaQueue with the given tag. 
+   *
+   * @param peelQueueTag Tag to access PeelQueue.
+   * @param peelQueue PeelQueue to add.
+   */
+  public void addPeelQueue(String peelQueueTag, PeelQueue peelQueue) {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: better to return a boolean to indicate if the operation is successful

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+    head.addPeelQueue(peelQueueTag, peelQueue);
+  }
+
+  /**
+   * Removes PeelQueue from the front element of the BananaQueue with the given tag. 
+   *
+   * @param peelQueueTag Tag to access PeelQueue.
+   */
+  public void removePeelQueue(String peelQueueTag) {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: return the removed PeelQueue

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+    head.removePeelQueue(peelQueueTag);
+  }
+
+  /**
+   * Adds BananaNode to the end of the queue and increments the size.
+   * 
+   * @param banana BananaNode to add
+   */
+  public void enqueueBanana(BananaNode banana) {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: return a boolean

File: capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    size++;
+  }
+  
+  /**
+   * Marks the front PeelNode in the queue as complete, decrements the size, and moves the front of 
+   * the queue to the next element (can be null).
+   */
+  public void dequeue() {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java]: consider returning the removed node

File: capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    foot = null;
+    size = 0;
+  }
+
+  /**
+   * Adds PeelNode to the end of the queue and increments the size.
+   * 
+   * @param peel PeelNode to add
+   */
+  public void enqueue(PeelNode peel) {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java]: consider returning a boolean

**Review by lupengscu — CHANGES_REQUESTED** [review]

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java (hunk 1 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java:27]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java (hunk 2 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java (hunk 3 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java (hunk 4 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java:21]: Should I add a comment to complete or write a better comment?

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 5 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java (hunk 6 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaNode.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 7 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 8 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: Choice between static addition of PeelNodes (creating a filled PeelQueue) and dynamic addition of PeelNodes (this case).

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 9 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 10 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 11 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java (hunk 12 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java (hunk 13 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java]: done

**Review by piercedwilson — APPROVED** [review]

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.util;
+
+public class BananaQueue {
```
Comment (sliang3) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java:48]: Can you provide a high level overview of how BananaQueue works as well?

File: capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    return size;
+  }
+  
+  /**
+   * Returns array representation of queue. [Currently deciding best approach]
+   *
+   * @return array representation of queue.
+   */
+  public PeelNode[] toArray() {
+    return null;
```
Comment (sliang3) [inline capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java:92]: Please add something like // TODO(ijelue): Figure out implementation.

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+   * Returns the number of BananaNodes in the BananaQueue.
+   * 
+   * @return number of BananaNodes in the BananaQueue.
+   */
+  public int getSize() {
+    return size;
+  }
+
+  /**
+   * Returns the number of PeelNodes in the PeelQueue of the associated tag. If not found,
```
Comment (sliang3) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: Not blocking for this review, but instead of using a number like this, you could consider using Optional (https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html).

File: capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+// limitations under the License.
+
+package com.google.sps;
+
+import org.junit.*;
+import static org.junit.Assert.*;
+import com.google.sps.util.*;
+import java.util.Objects;
+
+public class BananaQueueTest {
```
Comment (sliang3) [inline capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java:22]: For some of the longer tests, consider adding some comments to make each step more readable.

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+   * @return number of PeelNodes in the PeelQueue mapped to tag.
+   */
+  public int getPeelSize(String peelQueueTag) {
+    if(head == null || !head.peelQueueExists(peelQueueTag)) {
+      return -1;
+    }
+    return head.getPeelQueue(peelQueueTag).getSize();
+  }
+  
+  public void setSize(int size) {
```
Comment (sliang3) [inline capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: Do you need the setSize function?

File: capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+  
+  @Test
+  public void testEnqueuePeel() {
+    String peelQueueTag = "test";
```
Comment (sliang3) [inline capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java]: Seems like you can pull this out as a common variable.

Also, maybe you can add a test for upper case tags being converted to lower case.

Comment (ike-elue) [reply capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 18 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: I was thinking of needing it in the context of the progress model instead of the actual BananaQueue because I thought if a user wanted to revert the model, there'd be no way. However, this probably isn't needed/ruins the idea of the data structure. An alternative can be creating a whole new model based off the last one if a user wants to revert.

TL;DR: Makes sense in context of progress model but removed it because it doesn't fit BananaQueue

File: capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java (hunk 17 above)
Comment (ike-elue) [reply capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java:22]: done

File: capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java (hunk 15 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/PeelQueue.java:92]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 14 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java:48]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 16 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: I added Optional, but I wasn't too sure about the implementation. Is what I put, what you had in mind or should I just be returning the Optional object?

Comment (sliang3) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: I was thinking to return the Optional<Integer> and callers can check isPresent so that you can avoid using -1. Feel free to consider this at some other point though to get this submitted first.

File: capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java (hunk 17 above)
Comment (sliang3) [reply capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java:22]: nit: Please punctuate comments.

File: capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+
+public class BananaQueueTest {
+  
+  private final String peelQueueTag = "test";
+
+  @Test
+  public void testEnqueuePeel() {
+    BananaQueue bq = new BananaQueue();
+
+    // Try To add Peel without PeelQueue or BananaNode to attach to
```
Comment (sliang3) [inline capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java]: nit: lower case "To" here and below

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 18 above)
Comment (sliang3) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: Yeah I think this type of setter doesn't make as much sense for the BananaQueue data structure so if we decide we want to have a revert we can revisit how to do that later.

**Review by sliang3 — APPROVED** [review]

File: capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java (hunk 20 above)
Comment (ike-elue) [reply capstone/fitness/src/test/java/com/google/sps/BananaQueueTest.java]: done

File: capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java (hunk 16 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java]: done

**Review by lupengscu — APPROVED** [review]
